File: easy/relative_sort_array.go
package easy

import (
	"maps"
	"slices"
)

func RelativeSortArray(arr1, arr2 []int) []int {
	result := make([]int, len(arr1))
	freq := make(map[int]int)
	index := 0
	for _, num := range arr1 {
		freq[num]++
	}
	for _, num := range arr2 {
		count := freq[num]
		for count > 0 {
			result[index] = num
			count--
			index++
		}
		delete(freq, num)
	}

	for _, key := range slices.Sorted(maps.Keys(freq)) {
		count := freq[key]
		for count > 0 {
			count--
			result[index] = key
			index++
		}
	}
	return result
}

File: easy/relative_sort_array.py
from collections import Counter


def relative_sort_array(arr1: list[int], arr2: list[int]) -> list[int]:
    result = [0] * len(arr1)
    index = 0
    freq = Counter(arr1)
    for num in arr2:
        while freq[num]:
            result[index] = num
            freq[num] -= 1
            index += 1
        del freq[num]
    remainder = sorted(freq.items(), key=lambda x: x[0])
    for num, total in remainder:
        while total > 0:
            result[index] = num
            total -= 1
            index += 1
    return result
